add user.Info.GetExtra

deads2k opened the pull request:
I found myself wanting this field (or something like it), when trying to plumb the information about which scopes a particular token has.

Only the token authenticators have that information and I don't want tokens to leak past the authenticator.  I thought about extending the `authenticator.Token` interface to include scopes (`[]string`), but that felt a little specific for what I wanted to do.  I came up with this as an alternative.

It allows the token authenticator to fill in the information and authorizers already get handed the `user.Info`.  It means that implementors can choose to tie the layers together if they wish, using whatever data they think is best.  

@kubernetes/kube-iam

deads2k committed:
add user.Info.GetExtra

deads2k commented:
@erictune thoughts?

deads2k commented:
@kubernetes/kube-iam thoughts for or against?  The code is simple, we just need to agree on the concept.

liggitt commented:
I go back and forth... are scopes important enough to first-class? I would want to make it hard for anything doing authz checks to forget to include all the relevant information (for example, in a (Scoped?)SubjectAccessReview)

deads2k commented:
> On the specific vs generic accessor, I go back and forth... is the scope concept standard enough to give a first-class accessor?

I think having a first-class accessor that gives back information that must be interpreted in a standard way to a different pluggable component is a little weird.  Scopes are only as accurate as the authorizer using them and I'm afraid that creating a first-class accessor would give the impression that all authorizers should obey them.  I don't know that I'd expect that to be the case.

I'm also not convinced that we would agree on the shape of the those scopes.  Scopes aren't like usernames, they can't be opaque strings and provide the functionality that we want out of them.

erictune commented:
Can you elaborate on your use case of passing token scopes?

deads2k commented:
> Can you elaborate on your use case of passing token scopes?

1. UserA can create Pods
1. UserA tries to use SystemX
1. SystemX talks to an oauth server that issues tokens that are honored by a kube authenticator
1. SystemX asks for scopes: "see who i am" (proof of identity) and "selfsubjectaccessreview" (see what I can do)
1. UserA grants SystemX's client those scope.
1. oauth server gives Token1 to SystemX.
1. SystemX tries to do nefarious things with Token1 and create evil Pods
1. Scopes save the day because the kube authorizer being used respects the scopes set by the kube authenticator that say, "this token is only good for whoami and self-sar".

With SubjectAccessReviews (pull pending an update to client generation which has been opened), it becomes possible to build systems on top of kube for managing the system on behalf of particular users.  This will eventually require scopes for kube.

We hit this in openshift.

deads2k commented:
I should note that @liggitt brought up `SubjectAccessReview` needing to allow scopes as well and that has pretty much convinced me to make it a first class element.

@erictune Are you ok with making scopes first class on the `user.Info` object?

erictune commented:
Why, in step 3, would you give systemX a token that authenticates to a role which has more power that it should (step 7).  Why is "Extra" needed, when the token can just map to a suitable identity which has a role.

I mean, I get that it _could_ work the way you are proposing.  But I'd rather it didn't work that way, because:
- I'd really like each thing that interacts with the system to have a distinct identity.  An identity that can be audit-logged, rate-limited, used in debugging, and have its own distinct role (in multiple possible authorization systems).   An "anonymouns" oauth token that has authorization scopes, but whose identity is not known or is ignored (not mapped to a appropriately defined role), seems to go against that.  And that seems to be what is proposed here.
- so far, authentication and authorization are well separated, and I think it makes the system easier to understand and compose, and I'd like to keep that property.
- scope names on tokens is another namespace that needs to be managed, if it is an official part of Kubernetes.  And it seems confusing if you have token-scope-naming convention A, and then try to mix with authorization system B that has different names for things.  At the least, it is going to be hard to explain to users which components play together well.

erictune commented:
When you say "SubjectAccessReview needing to allow scopes", do you mean that you want "SubjectAccessReview" to be able to answer the question "Hey, I'm not going to tell you who I am, but suppose I showed up with a token with these scopes, would you let me do X?".

deads2k commented:
Updated godoc for `GetExtra`.  Also updated `SubjectAccessReview` to reflect the fact that it has to carry addition information to the authorizer for a decision.  That led to making it `map[string][]string` to be guaranteed of a clean serialization for someone like the webhook authorizer that may want to serialize the information for its use.

deads2k commented:
@k8s-bot test this issue #24508

deads2k commented:
@erictune bump

erictune commented:
Assigning to liggitt for detailed review.  Please somewhere document that (1) the Extra format is a json map, with key format undefined by K8s, (2) the admin needs to chose an Authenticator and Authorizer that understand the same Extra data, in order for it to work right, and (3) that Authenticator authors should chose keys that are unlikely to collide.

liggitt commented on pkg/auth/user/user.go:
```diff
@@ -27,6 +27,16 @@ type Info interface {
+	// GetExtra can contain any additional information that the authenticator
+	// thought was interesting.  One example would be scopes on a token.
+	// Keys in this map should be namespaced to the authenticator or
+	// authenticator/authorizer pair making use them.  For instance:
```
"use of them"

deads2k replied:
> "use of them"

done

liggitt commented:
nit, LGTM
